Modules/posixmodule.c

```diff
@@ -2588,12 +2588,43 @@ static int
 fill_time(PyObject *module, PyObject *v, int s_index, int f_index, int ns_index, time_t sec, unsigned long nsec)
 {
     assert(!PyErr_Occurred());
+#define S_TO_NS (1000000000LL)
+    assert(nsec < S_TO_NS);
+
+    if (f_index >= 0) {
+        PyObject *float_s = PyFloat_FromDouble(sec + 1e-9*nsec);
+        if (!float_s) {
+            return -1;
+        }
+        PyStructSequence_SET_ITEM(v, f_index, float_s);
+    }
+
+    /* 1677-09-21 00:12:44 to 2262-04-11 23:47:15 UTC inclusive */
+    if ((LLONG_MIN / S_TO_NS) <= sec && sec <= (LONG_MAX / S_TO_NS - 1)) {
+        if (s_index >= 0) {
+            PyObject *s = _PyLong_FromTime_t(sec);
+            if (!s) {
+                return -1;
+            }
+            PyStructSequence_SET_ITEM(v, s_index, s);
+        }
```

Comment:
This part does not require the guard `if ((LLONG_MIN / S_TO_NS) <= ...` above right? The code is repeated below. So can we move this part up and avoid duplication?

Reply:
In the fast path, when `s_index` is `-1`, we can skip allocating `*s`; when `s_index >= 0`, we allocate `*s` and unconditionally transfer ownership to the structseq.

In the slow path, we always allocate `*s` because we need to multiply it by `billion`, but we conditionally transfer ownership to the structseq based on `s_index` (and decref it when we don't).

That said, I see now that we can move the chunk you commented on above the fast path check if we have the slow path always allocate its own copy of `*s` so that the decref is safe.  We shouldn't ever take the slow path in practice, so that's fine.